pkg/middlewares/ratelimiter/redisrate/rate_test.go

```diff
@@ -0,0 +1,246 @@
+package redisrate_test
+
+import (
+	"context"
+	"testing"
+	"time"
+
+	"github.com/go-redis/redis_rate/v10"
+	"github.com/redis/go-redis/v9"
+	"github.com/stretchr/testify/require"
+)
+
+func rateLimiter() *redis_rate.Limiter {
```

Comment:
On Traefik, we avoid running unit tests against external components. When things needs to be tested against a real implementation we write integration tests (located in the `integration` directory).
So here I think it would be best faking the Redis store and execute the Lua script on an interpreter. 

I had great success in the past with `github.com/yuin/gopher-lua` and a hand rolled fake Redis implementation.
Something like this:

```go
type FakeRedisClient struct{
	script string
	keys   *ttlmap.TtlMap
}

func (m FakeRedisClient) EvalSha(ctx context.Context, _ string, keys []string, args ...interface{}) *redis.Cmd {
	state := lua.NewState()
	defer state.Close()

	tableKeys := state.NewTable()
	for _, key := range keys {
		tableKeys.Append(lua.LString(key))
	}
	state.SetGlobal("KEYS", tableKeys)

	tableArgv := state.NewTable()
	for _, arg := range args {
		tableArgv.Append(lua.LString(fmt.Sprint(arg)))
	}
	state.SetGlobal("ARGV", tableArgv)

	mod := state.SetFuncs(state.NewTable(), map[string]lua.LGFunction{
		"call": func(state *lua.LState) int {
			switch state.Get(1).String() {
			case "GET":
				key := state.Get(2).String()
				value, ok := m.keys.Get(key)
				if !ok {
					state.Push(lua.LNil)
				} else {
					state.Push(lua.LString(value.(string)))
				}
			case "SET":
				key := state.Get(2).String()
				value := state.Get(3).String()
				stringTTL := state.Get(5).String()
				ttl, _ := strconv.Atoi(stringTTL)
				_ = m.keys.Set(key, value, ttl)
			default:
				return 0
			}

			return 1
		},
	})
	state.SetGlobal("redis", mod)
	state.Push(mod)

	cmd := redis.NewCmd(ctx)
	if err := state.DoString("<script goes here>"); err != nil {
		cmd.SetErr(err)
		return cmd
	}

	result := state.Get(2)
	resultTable, ok := result.(*lua.LTable)
	if !ok {
		cmd.SetErr(errors.New("unexpected response type: " + result.String()))
		return cmd
	}

	var resultSlice []interface{}
	resultTable.ForEach(func(_ lua.LValue, value lua.LValue) {
		valueNbr, ok := value.(lua.LNumber)
		if !ok {
			valueStr, ok := value.(lua.LString)
			if !ok {
				cmd.SetErr(errors.New("unexpected response value type " + value.String()))
			}
			resultSlice = append(resultSlice, string(valueStr))
			return
		}

		resultSlice = append(resultSlice, int64(valueNbr))
	})

	cmd.SetVal(resultSlice)

	return cmd
}
```

Let me know if you need some help on the tests.

Reply:
Really thank you for this. It helps me a lot in my unit test